File: LeetCode/146. LRU Cache/solution.py
# class LRUCache:

#     def __init__(self, capacity: int):
#         self.cp = capacity
#         self.sz = 0
#         self.head = -2
#         self.tail = -1
#         self.mp = {}
#         self.mp[self.head] = {'val':-2, 'pre':-3, 'next':self.tail}
#         self.mp[self.tail] = {'val':-1, 'pre':self.head, 'next':0}

#     def get(self, key: int) -> int:
#         if key not in self.mp:
#             return -1
      
#         self.mp[self.mp[key]['pre']]['next'] = self.mp[key]['next']
#         self.mp[self.mp[key]['next']]['pre'] = self.mp[key]['pre']

#         pre = self.mp[self.tail]['pre']
#         self.mp[pre]['next'] = key
#         self.mp[key]['pre'] = pre
#         self.mp[key]['next'] = self.tail
#         self.mp[self.tail]['pre'] = key
  
#         return self.mp[key]['val']

#     def put(self, key: int, value: int) -> None:
#         # print(key)
#         if key in self.mp:
#             self.mp[self.mp[key]['pre']]['next'] = self.mp[key]['next']
#             self.mp[self.mp[key]['next']]['pre'] = self.mp[key]['pre']
#             del self.mp[key]
#             self.sz -= 1
#         elif self.sz == self.cp:
#             nxt = self.mp[self.head]['next']
#             self.mp[self.head]['next'] = self.mp[nxt]['next']
#             self.mp[self.mp[nxt]['next']]['pre'] = self.head
#             del self.mp[nxt]
#             self.sz -= 1
            
        
#         pre = self.mp[self.tail]['pre']
#         self.mp[pre]['next'] = key
#         self.mp[key] = {'val':value, 'pre':pre, 'next':self.tail}
#         self.mp[self.tail]['pre'] = key

#         self.sz += 1


class Node:
    def __init__(self, k, v):
        self.key = k
        self.val = v
        self.prev = None
        self.next = None

class LRUCache:
    def __init__(self, capacity):
        self.capacity = capacity
        self.dic = dict()
        self.head = Node(0, 0)
        self.tail = Node(0, 0)
        self.head.next = self.tail
        self.tail.prev = self.head

    def get(self, key):
        if key in self.dic:
            n = self.dic[key]
            self._remove(n)
            self._add(n)
            return n.val
        return -1

    def put(self, key, value):
        if key in self.dic:
            self._remove(self.dic[key])
        n = Node(key, value)
        self._add(n)
        self.dic[key] = n
        if len(self.dic) > self.capacity:
            n = self.head.next
            self._remove(n)
            del self.dic[n.key]

    def _remove(self, node):
        p = node.prev
        n = node.next
        p.next = n
        n.prev = p

    def _add(self, node):
        p = self.tail.prev
        p.next = node
        self.tail.prev = node
        node.prev = p
        node.next = self.tail
        
        


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: LeetCode/146. LRU Cache/solution.cpp
struct DLinkedNode {
    int key, value;
    DLinkedNode* prev;
    DLinkedNode* next;
    DLinkedNode(): key(0), value(0), prev(nullptr), next(nullptr) {}
    DLinkedNode(int _key, int _value): key(_key), value(_value), prev(nullptr), next(nullptr) {}
};

class LRUCache {
private:
    unordered_map<int, DLinkedNode*> cache;
    DLinkedNode* head;
    DLinkedNode* tail;
    int size;
    int capacity;

public:
    LRUCache(int _capacity): capacity(_capacity), size(0) {
        // 使用伪头部和伪尾部节点
        head = new DLinkedNode();
        tail = new DLinkedNode();
        head->next = tail;
        tail->prev = head;
    }
    
    int get(int key) {
        if (!cache.count(key)) {
            return -1;
        }
        // 如果 key 存在，先通过哈希表定位，再移到头部
        DLinkedNode* node = cache[key];
        moveToHead(node);
        return node->value;
    }
    
    void put(int key, int value) {
        if (!cache.count(key)) {
            // 如果 key 不存在，创建一个新的节点
            DLinkedNode* node = new DLinkedNode(key, value);
            // 添加进哈希表
            cache[key] = node;
            // 添加至双向链表的头部
            addToHead(node);
            ++size;
            if (size > capacity) {
                // 如果超出容量，删除双向链表的尾部节点
                DLinkedNode* removed = removeTail();
                // 删除哈希表中对应的项
                cache.erase(removed->key);
                // 防止内存泄漏
                delete removed;
                --size;
            }
        }
        else {
            // 如果 key 存在，先通过哈希表定位，再修改 value，并移到头部
            DLinkedNode* node = cache[key];
            node->value = value;
            moveToHead(node);
        }
    }

    void addToHead(DLinkedNode* node) {
        node->prev = head;
        node->next = head->next;
        head->next->prev = node;
        head->next = node;
    }
    
    void removeNode(DLinkedNode* node) {
        node->prev->next = node->next;
        node->next->prev = node->prev;
    }

    void moveToHead(DLinkedNode* node) {
        removeNode(node);
        addToHead(node);
    }

    DLinkedNode* removeTail() {
        DLinkedNode* node = tail->prev;
        removeNode(node);
        return node;
    }
};

作者：力扣官方题解
链接：https://leetcode.cn/problems/lru-cache/solutions/259678/lruhuan-cun-ji-zhi-by-leetcode-solution/
来源：力扣（LeetCode）
著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。

class LRUCache {
private:
    int capacity;
    list<int> ls;
    unordered_map<int,list<int>::iterator> mp;
    unordered_map<int,int> dict;
    void update(list<int>::iterator it, int key){
        ls.erase(it);
        ls.push_front(key);
        mp[key] = ls.begin();
        return;
    }
public:
    LRUCache(int capacity) {
        this->capacity = capacity;
    }
    
    int get(int key) {
        auto it = mp.find(key);
        if (it == mp.end()) return -1;
        auto mpit = mp[key];
        update(mpit, key);
        return dict[key];
    }
    
    void put(int key, int value) {
        if (get(key) == -1) {
            ls.push_front(key);
            mp[key] = ls.begin();
            dict[key] = value;
            capacity--;
            if (capacity < 0){
                capacity++;
                key = (ls.back());
                ls.pop_back();
                mp.erase(key);
                dict.erase(key);

            }
            return;
        }
        dict[key] = value;
        return;
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */


class LRUCache {

    int size;

    list<pair<int, int>> li;    //数据用双向链表保存

    unordered_map<int, list<pair<int, int>>::iterator> map; //value值是list的迭代器形式

public:

    LRUCache(int capacity) : size(capacity) {}  //初始化size大小

    

    int get(int key) {

        if(map.find(key) == map.end())  return -1;  //如果key值不存在,直接返回-1

        else

        {

            li.splice(li.begin(), li, map[key]);        //key存在需更新key对应的缓存到list的头部

            return map[key]->second;

        }

    }



    void put(int key, int value) {

        if(get(key) != -1)  map[key]->second = value;//key存在,调用get时已将缓存移到头部，再更新map

        else    

        {   //key不存在

            if(li.size() == size)  //缓存是否满了,删除最久未使用数据，map中也要删除

            {

                int delKey = li.back().first;

                li.pop_back();

                map.erase(delKey);

            }

            li.emplace_front(key, value); //如果key不存在，在头部加入新的缓存   

            map[key] = li.begin();

        }

    }

};

作者：zrita
链接：https://leetcode.cn/problems/lru-cache/solutions/848343/c-ji-jian-maplist-z-by-zrita-r0k0/
来源：力扣（LeetCode）
著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。